src/app/game-state/main-loop.service.ts

```diff
@@ -102,26 +110,39 @@ export class MainLoopService {
       const newTime = new Date().getTime();
       const timeDiff = newTime - this.lastTime;
       this.lastTime = newTime;
-      //this should be around 1, but may vary based on browser throttling
+      // this should be around 1, but may vary based on browser throttling
       let ticksPassed = timeDiff / TICK_INTERVAL_MS; 
+      const setTPS = this.getTPS(this.tickDivider) / 1000 * TICK_INTERVAL_MS;
+      const topTPS = this.getTPS(this.topDivider) / 1000 * TICK_INTERVAL_MS;
       if (this.pause) {
-        this.bankedTicks += ticksPassed/this.offlineDivider;
+        this.bankedTicks += ticksPassed * topTPS / this.offlineDivider; // offlineDivider currently either 10 or 2.
       } else {
-        ticksPassed *= this.getTPS(this.tickDivider)/1000*TICK_INTERVAL_MS;
+        let bankedCounter = 0;
         if (this.bankedTicks > 0 && this.useBankedTicks){
-          //using banked ticks makes time happen 10 times faster
-          ticksPassed *= 10;
-          this.bankedTicks -= timeDiff / TICK_INTERVAL_MS * 10;
+          bankedCounter = 11; // set to 10 + 1 for true / false
         }
+        this.bankedTicks += ticksPassed * (topTPS - setTPS) / this.offlineDivider; // Gain bankedticks for going slower than max speed.
 
-        this.tickCount += ticksPassed;
-        if (this.tickCount > 36500) {
-          //emergency lag prevention; this should never activate normally
-          this.tickCount = 36500;
-        }
-        while (!this.pause && this.tickCount > 0) {
-          this.tick();
-          this.tickCount--;
+        ticksPassed *= setTPS;
+        let tickTime = new Date().getTime();
+        if (bankedCounter) {
+          while (!this.pause && ticksPassed >= 1 && tickTime < TICK_INTERVAL_MS + newTime) {
+            this.tick();
+            if (this.bankedTicks >= 1 && bankedCounter > 1){ // Used here to avoid overusing earned bankedticks.
+              bankedCounter--
+              this.bankedTicks--
+            } else {
+              ticksPassed--;
+              bankedCounter = 11;
+            }
+            tickTime = new Date().getTime();
+          }
+        } else {
+          while (!this.pause && ticksPassed >= 1 && tickTime < TICK_INTERVAL_MS + newTime) {
+            this.tick();
+            ticksPassed--;
+            tickTime = new Date().getTime();
```

Comment:
Added a time comparison so that each tick checks if it's past the intended frametime.